apollo-router/src/plugins/telemetry/config_new/instruments.rs

```diff
@@ -1229,30 +1305,96 @@ where
             Increment::Duration(instant) | Increment::EventDuration(instant) => {
                 instant.elapsed().as_secs_f64()
             }
-            Increment::Custom(val) | Increment::EventCustom(val) => match val {
-                Some(incr) => incr as f64,
-                None => 0f64,
-            },
+            Increment::Custom(val) | Increment::EventCustom(val) | Increment::FieldCustom(val) => {
+                match val {
+                    Some(incr) => incr as f64,
+                    None => 0f64,
+                }
+            }
         };
 
         if let Some(counter) = inner.counter.take() {
             counter.add(increment, &attrs);
         }
     }
+
+    fn on_response_field(&self, typed_value: &TypedValue, ctx: &Context) {
+        let mut inner = self.inner.lock();
+        if !inner.condition.evaluate_response_field(typed_value, ctx) {
+            return;
+        }
+
+        // Response field may be called multiple times so we don't extend inner.attributes
+        let mut attrs = inner.attributes.clone();
+        if let Some(selectors) = inner.selectors.as_ref() {
+            attrs.extend(
+                selectors
+                    .on_response_field(typed_value, ctx)
+                    .into_iter()
+                    .collect::<Vec<_>>(),
+            );
+        }
+
+        if let Some(selected_value) = inner
+            .selector
+            .as_ref()
+            .and_then(|s| s.on_response_field(typed_value, ctx))
+        {
+            let new_incr = match &inner.increment {
+                Increment::FieldCustom(None) => {
+                    Increment::FieldCustom(selected_value.as_str().parse::<i64>().ok())
+                }
+                Increment::Custom(None) => {
+                    Increment::FieldCustom(selected_value.as_str().parse::<i64>().ok())
+                }
+                other => {
+                    failfast_error!("this is a bug and should not happen, the increment should only be Custom or FieldCustom, please open an issue: {other:?}");
+                    return;
+                }
+            };
+            inner.increment = new_incr;
+        }
+
+        let increment: Option<f64> = match &mut inner.increment {
+            Increment::FieldUnit => Some(1f64),
+            Increment::FieldCustom(val) => {
+                let incr = val.map(|incr| incr as f64);
+                // Set it to None again
+                *val = None;
+                incr
+            }
+            Increment::Unit
+            | Increment::Duration(_)
+            | Increment::Custom(_)
+            | Increment::EventDuration(_)
+            | Increment::EventCustom(_)
+            | Increment::EventUnit => {
+                // Nothing to do because we're incrementing on fields
+                return;
+            }
+        };
+
+        if let (Some(counter), Some(increment)) = (&inner.counter, increment) {
```

Comment:
Just to make sure, I think we have to use the same mechanism than for `response_event` here with the `incremented` flag. It's useful when the instrument is dropped. Could you make sure we don't need to set it to true somewhere here ?

Reply:
[90135b6](https://github.com/apollographql/router/pull/5215/commits/90135b6d74d14dbc0091060772b0e78eddcc72d3)